rmf_fleet_adapter/src/rmf_fleet_adapter/TaskManager.cpp

```diff
@@ -1678,49 +1736,130 @@ void TaskManager::_publish_task_state()
+  const rmf_task::Parameters& parameters)
+{
+  const auto info = pending.request()->description()->generate_info(
+    std::move(expected_state), parameters);
+
+  nlohmann::json pending_json;
+  const auto& booking = *pending.request()->booking();
+  copy_booking_data(pending_json["booking"], booking);
+
+  pending_json["category"] = info.category;
+  pending_json["detail"] = info.detail;
+
+  pending_json["unix_millis_start_time"] =
+    to_millis(pending.deployment_time().time_since_epoch()).count();
+
+  if (pending.finish_state().time())
+  {
+    pending_json["unix_millis_finish_time"] =
+      to_millis(pending.finish_state().time()->time_since_epoch()).count();
+
+    const auto estimate =
+      pending.finish_state().time().value() - pending.deployment_time();
+    pending_json["original_estimate_millis"] =
+      std::max(0l, to_millis(estimate).count());
+  }
+  copy_assignment(pending_json["assigned_to"], *_context);
+  pending_json["status"] = "queued";
+
+  auto task_state_update = _task_state_update_json;
+  task_state_update["data"] = pending_json;
+
+  static const auto validator =
+    _make_validator(rmf_api_msgs::schemas::task_state_update);
+
+  _validate_and_publish_websocket(task_state_update, validator);
+
+  return pending.finish_state();
+}
+
 //==============================================================================
 void TaskManager::_publish_task_queue()
 {
   rmf_task::State expected_state = _context->current_task_end_state();
   const auto& parameters = *_context->task_parameters();
-  static const auto validator =
-    _make_validator(rmf_api_msgs::schemas::task_state_update);
+
+  for (const auto& pending : _direct_queue)
+  {
+    expected_state = _publish_pending_task(
+      pending.assignment, std::move(expected_state), parameters);
+  }
 
   for (const auto& pending : _queue)
   {
-    const auto info = pending.request()->description()->generate_info(
-      expected_state, parameters);
+    expected_state = _publish_pending_task(
+      pending, std::move(expected_state), parameters);
+  }
+}
 
-    nlohmann::json pending_json;
-    const auto& booking = *pending.request()->booking();
-    copy_booking_data(pending_json["booking"], booking);
+//==============================================================================
+void TaskManager::_publish_canceled_pending_task(
+  const Assignment& pending,
+  std::vector<std::string> labels)
+{
+  nlohmann::json pending_json;
+  const auto& booking = *pending.request()->booking();
+  copy_booking_data(pending_json["booking"], booking);
 
-    pending_json["category"] = info.category;
-    pending_json["detail"] = info.detail;
+  pending_json["unix_millis_start_time"] =
+    to_millis(pending.deployment_time().time_since_epoch()).count();
 
-    pending_json["unix_millis_start_time"] =
-      to_millis(pending.deployment_time().time_since_epoch()).count();
+  copy_assignment(pending_json["assigned_to"], *_context);
+  pending_json["status"] = "canceled";
```

Comment:
Got confused by this for a while and thought it was a typo, we might need to converge to single or double L at some point, I found multiple locations where we use double Ls, https://github.com/open-rmf/rmf_api_msgs/blob/main/rmf_api_msgs/schemas/task_state.json#L51.

This should not block merging this PR though

Reply:
Yeah, the inconsistency between the number of `l`s in `canceled` versus `cancellation` probably creates a high risk of typos, but I chose those spellings as they're the more common spellings in American English, which is usually preferred over British spellings in the global software industry. I'm open to revising the API if people anticipate that it will cause problems.